### PR TITLE
Bump to Doctrine CS 4.0

### PR DESCRIPTION
**Pull request by carusogabriel**

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -13,9 +13,11 @@ php:
 
 before_install:
   - mv ~/.phpenv/versions/$(phpenv version-name)/etc/conf.d/xdebug.ini{,.disabled} || echo "xdebug not available"
-  - travis_retry composer self-update
+  - composer self-update
```
Comment (Majkl578) [inline .travis.yml]: Why did you remove this?

Comment (carusogabriel) [reply .travis.yml]: Just copy-paste from ORM. Why do we have it here?

Comment (Majkl578) [reply .travis.yml]: It's simply _retry if fails, up to 3 times_. Would probably add it there rather than remove it here.
Good for network operations since Travis can be unstable randomly.

Comment (carusogabriel) [reply .travis.yml]: Sure, gonna do that though.

**Comment by carusogabriel** [comment]
Conflicts solved, should be RTM :smile:

File: lib/Doctrine/Common/Inflector/Inflector.php [hunk 2]
```diff
@@ -1,5 +1,7 @@
 <?php
 
+declare(strict_types=1);
```
Comment (Majkl578) [inline lib/Doctrine/Common/Inflector/Inflector.php]: Not possible in stable branch.

Comment (carusogabriel) [reply lib/Doctrine/Common/Inflector/Inflector.php]: Why? :(

Comment (carusogabriel) [reply lib/Doctrine/Common/Inflector/Inflector.php]: Yeah, [we can't](https://secure.php.net/manual/en/functions.arguments.php#functions.arguments.type-declaration.strict).

Reverted, now, RTM :)

**Comment by carusogabriel** [comment]
I've also added PHPStan at max level in this PR :smile:

File: composer.json [hunk 3]
```diff
@@ -16,8 +16,9 @@
         "php": "^7.1"
     },
     "require-dev": {
-        "doctrine/coding-standard": "^3.0",
-        "phpunit/phpunit": "^7.0"
+        "doctrine/coding-standard": "^4.0",
+        "phpunit/phpunit": "^7.0",
+        "phpstan/phpstan": "^0.9.2"
```
Comment (Majkl578) [inline composer.json]: Please use phpstan/phpstan-shim.

Comment (Majkl578) [reply composer.json]: And reorder aplhabetically.

File: phpcs.xml.dist [hunk 4]
```diff
@@ -14,7 +14,5 @@
 
     <rule ref="Doctrine">
         <exclude name="SlevomatCodingStandard.TypeHints.DeclareStrictTypes"/>
-        <exclude name="SlevomatCodingStandard.TypeHints.TypeHintDeclaration.MissingParameterTypeHint"/>
-        <exclude name="SlevomatCodingStandard.TypeHints.TypeHintDeclaration.MissingReturnTypeHint"/>
```
Comment (Majkl578) [inline phpcs.xml.dist]: Not possible, keep these.

Comment (Majkl578) [reply phpcs.xml.dist]: Hmm actually looks ok here, inflector is already typed. :)

File: composer.json (hunk 3 above)
Comment (carusogabriel) [reply composer.json]: Done :+1:

File: lib/Doctrine/Common/Inflector/Inflector.php [hunk 5]
```diff
@@ -31,7 +31,7 @@ class Inflector
     /**
      * Plural inflector rules.
      *
-     * @var string[][]
+     * @var string|string[][]
```
Comment (Majkl578) [inline lib/Doctrine/Common/Inflector/Inflector.php]: I think this is definitely wrong. It is array of arrays, not a single string... Why was this done? Did PHPStan suggest that?

Comment (carusogabriel) [reply lib/Doctrine/Common/Inflector/Inflector.php]: The error log:
```
 ------ ------------------------------------------------------ 
  Line   lib/Doctrine/Common/Inflector/Inflector.php           
 ------ ------------------------------------------------------ 
  489    Array (array<string>) does not accept array<string>.  
  497    Array (array<array<string>>) does not accept string.  
  498    Array (array<array<string>>) does not accept string.  
  550    Array (array<array<string>>) does not accept string.  
  551    Array (array<array<string>>) does not accept string.  
 ------ ------------------------------------------------------
```

**Comment by carusogabriel** [comment]
RTM? :)

File: lib/Doctrine/Common/Inflector/Inflector.php (hunk 5 above)
Comment (alcaeus) [reply lib/Doctrine/Common/Inflector/Inflector.php]: Looks like PHPStan has some hiccups here. Might be worth reporting upstream, in the meantime I'd be fine merging this since the members are private. Maybe we can add a comment explaining it?
